### ISSUE TITLE
Don't use HTTP 503

### ISSUE THREAD

**Issue by rmannibucau** [eclipse/microprofile-health#129]
this status is misleading from a client point of view, a /health endpoint should return a 200 to say "we did the test and it failed" instead of "the server failed to do the test"

**Comment by pilhuhn** [comment]
In case of failure it should not return a 200 code. Please see https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#define-a-liveness-http-request

> Any code greater than or equal to 200 and less than 400 indicates success. Any other code indicates failure.

One could argue if the endpoint should return a 509, 503 depending on what failed or just a 500, which may even be better than a 503 which means "temporarily unavailable", but a 200 is imo not the right thing.

**Comment by rmannibucau** [comment]
@pilhuhn ok, let me rephrase, anything in <300 is ok but 3xx, 4xx and 5xx are misleading from a client point of view since it is about the endpoint status and not the content of the payload of the status.

If you want such behavior I'd ask to provide a query param to handle it: /health?statusOnFailure=500 which is trivial to impl and provide a way to match all clients.

**Comment by pilhuhn** [comment]
@rmannibucau The clients for the `/health` endpoint are automatic processes, that check if the running container is able to do its work or not. In the later case the container will be killed and a new one will be instantiated.
For that checker process it does not really matter if if can't reach `/health`, if the endpoint does not react or if the endpoint indicates that the check procedures reported "unhealthy". In all those cases it should consider the container as unhealthy and kill it.

For a human as client, the situation may indeed be different, but humans are not really the target here.
So the default should be to fit what Kubernetes does. And an override for humans could work differently.

**Comment by rmannibucau** [comment]
@pilhuhn I know. let me try to develop my point: human can always check other things and it is not friendly to have a 5xx but it is usable. Automatic process can be plain stupid and just want a toggle (200=OK, 500=KO) or something finer to know if the endpoint is broken or if the server is broken or if some service is broken. It is not rare nowdays (since 5 years actually) to have automatic client checkers able to parse JSON and therefore if I have a MySQL check and a Mongo check,  to say that MySQL failure will result in my system being read and Mongo will just make it yellow. However if I get a 500 it must mean the endpoint is broken and the system is therefore red since even the healthcheck doesnt respond anymore.

> So the default should be to fit what Kubernetes does.

I don't think so, the default should fit the mainstream in the industry and not a particular technology. Health checks is a very old and spread solution and even in k8s clusters, users often rely on another check technology cause k8s one is very low level and doesn't help enough ops to work ("when we start to have one failure the whole cluster is often failing").

See it this way: what if we add this statusOnFailure param? k8s still works since you just change the url and all other cases are enabled, isnt it a win-win?

**Comment by struberg** [comment]
FTR: I agree with Romain. It's not always 'we are all f*cked up' but rather: "meh I feel like puking, gimme a few minutes'.

In production I've rarely seen a problem which goes away with a reboot. Yes there are mem leaks etc. But they only get worse with the bazooka way of fixing them. You know, this solution is called Cold Reboot And Pray - short CRAP ;)

Any robot like client has also a chance to evaluate the content of the reply. So there is no need to re-use an InternalServerError (or worse UserError) HTTP code. It's really bad practice.

**Comment by rmannibucau** [comment]
Well if we add a query param to current endpoint we match all cases and it avoids 2 calls for the same "information" and this is needed for automated health checks and not only human checks (that's what I use ATM, with json parsing)

**Comment by derekm** [comment]
While I agree with @rmannibucau & @struberg in principle, it seems like cloud health checks have already set the precedent that health endpoints use the status codes in very special ways.

Even Azure agrees with Kubernetes on this:
https://docs.microsoft.com/en-us/azure/architecture/patterns/health-endpoint-monitoring

> Most existing tools and frameworks look only at the HTTP status code that the endpoint returns.

And then they go on to use 400 and 500 status codes the same way Kubernetes does.

**Comment by rmannibucau** [comment]
And companies healthchecks don't all do that so we are exactly in the case where it must be configurable from the client side IMHO.

**Comment by derekm** [comment]
@cescoffier -- I'm seeing requests for "hints" and more "human readable" output here and elsewhere. Why is the existing `data` section (called `details` in Spring) not sufficient? I put my human readable content in the `data` sections of my checks. Am I missing something?

**Comment by antoinesd** [comment]
won't fix

**Comment by rmannibucau** [comment]
Why? There is not only k8s and this does not work in alternatives or company specific solutions in general requiring to rewrite checks to fail a success :(
